python/cudf/cudf/testing/testing.py

```diff
@@ -201,31 +206,49 @@ def assert_column_equal(
     ):
         left = left.astype(left.categories.dtype)
         right = right.astype(right.categories.dtype)
-
     columns_equal = False
-    try:
-        columns_equal = (
-            (
-                cp.all(left.isnull().values == right.isnull().values)
-                and cp.allclose(
+    if left.size == right.size == 0:
+        columns_equal = True
+    elif not (
+        (is_string_dtype(left) and is_numeric_dtype(right))
+        or (is_numeric_dtype(left) and is_string_dtype(right))
+    ):
+        try:
+            # nulls must be in the same places for all dtypes
+            nulls_equal = cp.all(left.isnull().values == right.isnull().values)
+
+            if not check_exact and is_numeric_dtype(left):
+                # non-null values must be the same
+                values_equal = cp.allclose(
                     left[left.isnull().unary_operator("not")].values,
                     right[right.isnull().unary_operator("not")].values,
                 )
-            )
-            if not check_exact and is_numeric_dtype(left)
-            else left.equals(right)
-        )
-    except TypeError as e:
-        if str(e) != "Categoricals can only compare with the same type":
-            raise e
-        if is_categorical_dtype(left) and is_categorical_dtype(right):
-            left = left.astype(left.categories.dtype)
-            right = right.astype(right.categories.dtype)
+                if not left.dtype.kind == right.dtype.kind == "f":
+                    columns_equal = nulls_equal and values_equal
+                else:
+                    # nans must be the same for float types
+                    nans_equal = cp.all(
+                        is_nan(left).values == is_nan(right).values
+                    )
+                    columns_equal = nulls_equal and values_equal and nans_equal
+            else:
+                columns_equal = left.equals(right)
+        except TypeError as e:
+            if str(e) != "Categoricals can only compare with the same type":
+                raise e
+            if is_categorical_dtype(left) and is_categorical_dtype(right):
+                left = left.astype(left.categories.dtype)
+                right = right.astype(right.categories.dtype)
```

Comment:
These if elses become a little cleaner if you eschew the `nulls_equal` and `nans_equal` variables in favor of just assigning back to `columns_equal`. I can't suggest because there's intervening unchanged lines, but basically just do
```
columns_equal = cp.all(...) # Replace assigning to nulls_equal
if columns_equal and not check_exact...:
    columns_equal = cp.allclose...
    if columns_equal and (left.dtype.kind == right.dtype.kind == "f"): # Note that I've inverted the second condition
        columns_equal = cp.all(...)
```
This also removes unnecessary checks if one of the previous ones has already failed.

Reply:
This worked great, but also requires a quick update to the categorical handling below.